### PR TITLE
Finalizer's and Temporary

### PR DESCRIPTION
**Pull request by nhibberd**
Making RIO Finalizer's safe to run and cleaning up Temporary

- [x] poacher
- [x] saws
- [x] notion
- [x] ivory

### CODE REVIEW

File: mundane-io/src/main/scala/com/ambiata/mundane/io/LocalTemporary.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
+package com.ambiata.mundane.io
+
+import com.ambiata.mundane.control._
+import com.ambiata.mundane.io.Temporary._
+import java.util.concurrent.atomic.AtomicInteger
+import scalaz._, Scalaz._
+
+case class LocalTemporary(seed: String) {
+  private var step: AtomicInteger = new AtomicInteger(0)
```
Comment (charleso) [inline mundane-io/src/main/scala/com/ambiata/mundane/io/LocalTemporary.scala]: `val`? /supernit

File: mundane-io/src/test/scala/com/ambiata/mundane/io/LocalTemporarySpec.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+   no conflicts                                      $conflicts
+
+"""
+  /** Testing Temporary clean up finalizers */
+  def file = prop((data: String, tmp: LocalTemporary) => for {
+    f <- tmp.fileWithContent(data)
+    e <- Files.exists(f)
+    _ <- RIO.unsafeFlushFinalizers
+    z <- Files.exists(f)
+  } yield e -> z ==== true -> false)
```
Comment (charleso) [inline mundane-io/src/test/scala/com/ambiata/mundane/io/LocalTemporarySpec.scala]: I'm not 100% sure, but I think these tests are being ignored. You're not implementing `ThrownExceptions` (which you shouldn't), and you're not using `ResultTMatcher`. Let me know if I'm missing something.

Comment (nhibberd) [inline mundane-io/src/test/scala/com/ambiata/mundane/io/LocalTemporarySpec.scala]: implicit in RIOMatcher for AsResult and Prop

File: mundane-control/src/main/scala/com/ambiata/mundane/control/RIO.scala [hunk 3]
```diff
@@ -72,6 +74,17 @@ object RIO {
   def addFinalizer(f: Finalizer): RIO[Unit] =
     RIO(finalizers => { finalizers.add(f); ResultT.unit })
 
+  def unsafeFlushFinalizers: RIO[Unit] = RIO(x => {
+    ResultT.safe[IO, Unit](x.asScala.toList.foreach(_.run.unsafePerformIO match {
```
Comment (charleso) [inline mundane-control/src/main/scala/com/ambiata/mundane/control/RIO.scala]: Would we want to run all the finalizers even if one of them fails (and then throw an exception after)?

**Pull request merged**